### PR TITLE
Added transformer example

### PR DESCRIPTION
**Pull request by apoorvnandan**
The example shows how to implement a transformer block as a Keras layer and use it for text classification on IMDB data. 

The model trains faster than an LSTM based classifier with a similar number of parameters and gives slightly better results.

- The transformer shown in the example is a small one; 2 attention heads, and small dimensions for projection heads and FFN. It achieves validation accuracy of 87% on IMDB in less than a minute on CPU.
- There is not a whole lot of description about the multi-head self-attention part, as I feel like that would make this into a detailed tutorial. Just implemented a simple version of the transformer block which can be used in place of an LSTM or GRU layer.  
(More description can be added if the maintainers feel like it is needed)
- Positional embedding are implemented using a standard `keras.layers.Embedding` layer, which according to the paper produced nearly identical results to the sinusoidal version. 
- For the classification head, I am taking the mean of transformer outputs across all timesteps. This has turned out to be a good practice even with larger transformer encoders like BERT.

### CODE REVIEW

File: examples/nlp/text_classification_with_transformer.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+Date created: 2020/05/10
+Last modified: 2020/05/10
+Description: Implement transformer block as a Keras layer and use it for text classification.
+"""
+"""
+## Setup
+"""
+
+import tensorflow as tf
+from tensorflow import keras
```
Comment (fchollet) [inline examples/nlp/text_classification_with_transformer.py:13]: Add `from tensorflow.keras import layers`, and use `layers.` whenever you use a layer object in your code

File: examples/nlp/text_classification_with_transformer.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+## Implement multi head self attention as a Keras layer
+"""
+
+
+class MultiHeadSelfAttention(keras.layers.Layer):
+    def __init__(self, embed_dim, num_heads=8):
+        super(MultiHeadSelfAttention, self).__init__()
+        self.embed_dim = embed_dim
+        self.num_heads = num_heads
+        assert (
```
Comment (fchollet) [inline examples/nlp/text_classification_with_transformer.py:25]: Do not validate user-provided arguments with `assert`, instead, raise `ValueError`. In the message, print what was provided and what was expected.

File: examples/nlp/text_classification_with_transformer.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        scaled_score = score / tf.math.sqrt(dk)
+        weights = tf.nn.softmax(scaled_score, axis=-1)
+        output = tf.matmul(weights, v)
+        return output, weights
+
+    def separate_heads(self, x, batch_size):
+        x = tf.reshape(x, (batch_size, -1, self.num_heads, self.projection_dim))
+        return tf.transpose(x, perm=[0, 2, 1, 3])
+
+    def call(self, x):
```
Comment (fchollet) [inline examples/nlp/text_classification_with_transformer.py:46]: Use `inputs` in the signature instead of `x` (you can use `x` in the code of the `call` method), to be consistent with the standard `call` signature

File: examples/nlp/text_classification_with_transformer.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        assert (
+            embed_dim % num_heads == 0
+        ), "embedding dimension not divisible by num heads"
+        self.projection_dim = embed_dim // num_heads
+        self.wq = keras.layers.Dense(embed_dim)
+        self.wk = keras.layers.Dense(embed_dim)
+        self.wv = keras.layers.Dense(embed_dim)
+        self.combine_heads = keras.layers.Dense(embed_dim)
+
+    def attention(self, q, k, v):
```
Comment (fchollet) [inline examples/nlp/text_classification_with_transformer.py:34]: Prefer using fully spelled out argument names, e.g. `query`, `key`, `value`.

File: examples/nlp/text_classification_with_transformer.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+class MultiHeadSelfAttention(keras.layers.Layer):
+    def __init__(self, embed_dim, num_heads=8):
+        super(MultiHeadSelfAttention, self).__init__()
+        self.embed_dim = embed_dim
+        self.num_heads = num_heads
+        assert (
+            embed_dim % num_heads == 0
+        ), "embedding dimension not divisible by num heads"
+        self.projection_dim = embed_dim // num_heads
+        self.wq = keras.layers.Dense(embed_dim)
```
Comment (fchollet) [inline examples/nlp/text_classification_with_transformer.py:29]: Prefer using fully spelled out argument names, e.g. `query_dense` or `query_proj`

File: examples/nlp/text_classification_with_transformer.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        self.dropout2 = tf.keras.layers.Dropout(rate)
+
+    def call(self, x, training):
+        attn_output = self.att(x)
+        attn_output = self.dropout1(attn_output, training=training)
+        out1 = self.layernorm1(x + attn_output)
+
+        ffn_output = self.ffn(out1)
+        ffn_output = self.dropout2(ffn_output, training=training)
+        out2 = self.layernorm2(out1 + ffn_output)
```
Comment (fchollet) [inline examples/nlp/text_classification_with_transformer.py:104]: You can put the `return` on this line

File: examples/nlp/text_classification_with_transformer.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+
+"""
+## Implement embedding layer
+
+Two seperate embedding layers, one for tokens, one for token index (positions).
+"""
+
+
+class EmbeddingLayer(keras.layers.Layer):
```
Comment (fchollet) [inline examples/nlp/text_classification_with_transformer.py:116]: Use a more specific name, since `Embedding` is already a layer. E.g. `TokenAndPositionEmbedding`

File: examples/nlp/text_classification_with_transformer.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+
+"""
+## Create classifier model using transformer layer
+
+Transformer layer outputs one vector for each time step of your input sequence. Here, we take the mean across all time steps and build a two layered feed forward network on top of it.
+"""
+
+
+class TransformerClassifier(tf.keras.Model):
```
Comment (fchollet) [inline examples/nlp/text_classification_with_transformer.py:139]: Prefer using the Functional API to implement this model. Subclassing is more verbose and restrict usable features at no gain in this case.

File: examples/nlp/text_classification_with_transformer.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+        x = self.dropout2(x, training=training)
+        x = self.head(x)
+        return x
+
+
+"""
+## Download and prepare dataset
+"""
+
+max_features = 6000  # Only consider the top 20k words
```
Comment (fchollet) [inline examples/nlp/text_classification_with_transformer.py:164]: 6000 in this case, not 20k. I recommend using at least 15k though. 6k is too low.

File: examples/nlp/text_classification_with_transformer.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+        )  # (batch_size, seq_len, embed_dim)
+        return output
+
+
+"""
+## Implement transformer block as a layer
+"""
+
+
+class TransformerLayer(tf.keras.layers.Layer):
```
Comment (fchollet) [inline examples/nlp/text_classification_with_transformer.py:79]: `TransformerBlock` would more appropriate. Also in general we don't put `Layer` in layer names.

File: examples/nlp/text_classification_with_transformer.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+        positions = tf.range(start=0, limit=maxlen, delta=1)
+        positions = self.pos_emb(positions)
+        x = self.token_emb(x)
+        return x + positions
+
+
+"""
+## Create classifier model using transformer layer
+
+Transformer layer outputs one vector for each time step of your input sequence. Here, we take the mean across all time steps and build a two layered feed forward network on top of it.
```
Comment (fchollet) [inline examples/nlp/text_classification_with_transformer.py:135]: Break up text lines so that they are no longer than 90 char. This helps with version control and code readability.

**Comment by apoorvnandan** [comment]
Thank you for the comments. I will go through them and make appropriate changes.

**Comment by apoorvnandan** [comment]
Please review the updated code.

**Comment by apoorvnandan** [comment]
Added md and ipynb files from `autogen.py`.
